### PR TITLE
Add sd_notify() call to custodia server

### PR DESCRIPTION
**Pull request by tiran**
The Custodia server is now able to notify systemd that it is ready to
handle requests. systemd.daemon.notify("READY=1") is called after the
listening socket has been created and before the HTTP server starts its
server loop.

Custodia depends on python-systemd or python3-systemd for notification.
If no systemd bindings are available or systemd isn't running,
sd_notify() is not called.

Signed-off-by: Christian Heimes <cheimes@redhat.com>

### CODE REVIEW

File: custodia/httpd/server.py [hunk 1]
```diff
@@ -21,6 +21,14 @@
     from socketserver import ForkingTCPServer
     from urllib.parse import urlparse, parse_qs, unquote
 
+try:
+    # pylint: disable=import-error
+    from systemd.daemon import notify as sd_notify
+    from systemd.daemon import booted as systemd_booted
```
Comment (simo5) [inline custodia/httpd/server.py]: can we use a consistent naming here ?
either sd_notify, sd_booted or systemd_notify, systemd_booted work for me.

File: custodia/httpd/server.py [hunk 2]
```diff
@@ -414,4 +423,7 @@ def get_socket(self):
         return (self.httpd.socket, self.httpd.socket_file)
 
     def serve(self):
+        if sd_notify is not None and systemd_booted():
+            logger.debug('sd_notify("READY=1")')
```
Comment (simo5) [inline custodia/httpd/server.py]: What's the point of this log statement ?
I do not think we really care to log a functional detail of the startup process, systemd is already loggin that for us.

**Comment by tiran** [comment]
I have addressed both comments.

**Comment by simo5** [comment]
Build fails, please fix the ImportEr_R_or issue.

**Comment by tiran** [comment]
For FreeIPA downstream custodia.service needs to be changed to allow and use notifications:
```
Type=notify
NotifyAccess=main
```

custodia.spec needs require python-systemd / python3-systemd, too.

**Comment by simo5** [comment]
It'd be nice if you could submit a patch to freeipa-devel for the FreeIPA remarks

**Pull request closed without merging**